packages/eslint-scope/lib/referencer.js

```diff
@@ -649,6 +649,50 @@ class Referencer extends esrecurse.Visitor {
 
         // do nothing.
     }
+
+    JSXIdentifier(node) {
+
+        if (this.scopeManager.__isJSXEnabled()) {
+            this.currentScope().__referencing(node);
+        }
+    }
+
+    JSXMemberExpression(node) {
+        this.visit(node.object);
+    }
+
+    JSXElement(node) {
+        if (this.scopeManager.__isJSXEnabled()) {
+            this.visit(node.openingElement);
+            node.children.forEach(this.visit, this);
+        } else {
+            this.visitChildren(node);
+        }
+    }
+
+    JSXOpeningElement(node) {
+        if (this.scopeManager.__isJSXEnabled()) {
+
+            const nameNode = node.name;
+
+            // we only want to visit JSXIdentifier nodes if they are capitalized
+            if (nameNode.type !== "JSXIdentifier" || nameNode.name[0].toUpperCase() === nameNode.name[0]) {
```

Comment:
There's also [`JSXNamespacedName`](https://facebook.github.io/jsx/#prod-JSXNamespacedName) syntax, e.g., `<A:B/>`. Per the current implementation, both `A` and `B` would be treated as references. I'm not sure if that's correct or incorrect as I'm unsure about the semantics. In React, I don't think this syntax has any meaning, and `react/jsx-uses-vars` just [skips it](https://github.com/jsx-eslint/eslint-plugin-react/blob/e6b5b41191690ee166d0cca1e9db27092b910f03/lib/rules/jsx-uses-vars.js#L35-L38).

Reply:
Can you point me to an example of how that is used? Googling "JSX namespace" doesn't bring up anything.

Reply:
After asking on social media, confirmed that React doesn't use this at all and the one reference someone shared with me (https://github.com/appsemble/appsemble/blob/3fbb56aa1b47e8812e7c56ae6dfa6d446c6503f6/packages/server/controllers/main/apps/saml/assertAppSamlConsumerService.test.tsx#L30..L157) did not treat the element as a component reference even though it was capitalized. So, ensured that these are just ignored.

Reply:
Yeah, I couldn't find any uses of this syntax in any uses of JSX. Seems best to just ignore it and make sure that neither of the identifiers used in this syntax are treated as references.

Reply:
Yup, that's how I implemented it.